### ISSUE TITLE
Validate variable dtype before restoring checkpoint

### ISSUE THREAD

**Issue by k-w-w** [tensorflow/tensorflow#20487]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: [Yes](https://stackoverflow.com/questions/51137417/wrong-output-for-restored-variable-in-tensorflow-graph)
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: 1.10.0-dev20180620
- **Python version**:  2.7
- **Bazel version (if compiling from source)**: N/A
- **GCC/Compiler version (if compiling from source)**: N/A
- **CUDA/cuDNN version**: 9.0 / 7
- **GPU model and memory**: GeForce GTX 1080 / 8 Gb
- **Exact command to reproduce**: Run the two scripts below in "Source code/logs"

### Describe the problem
A warning/error should be raised when a variable is loaded with a different dtype as that which was saved. 

In the code below, the variable `global_step` is saved as a `tf.int32`, but treated as a `tf.float32` when restored. When this happens, the `global_step` variable appears to have a value of 7e-45 instead of the expected value of 5. 

### Source code / logs
 [Link to Stackoverflow Post with the code/question](https://stackoverflow.com/questions/51137417/wrong-output-for-restored-variable-in-tensorflow-graph)

[Original post copied here]
**Script that saves the checkpoints:**

    import tensorflow as tf

    a = tf.Variable(3.0, name='a')
    b = tf.Variable(5.0, name='b')
    
    b = tf.assign_add(b, a)
    
    n_steps = 5
    
    global_step = tf.Variable(0, name='global_step', trainable=False)
    
    saver = tf.train.Saver()
    
    with tf.Session() as sess:
    
        sess.run(tf.global_variables_initializer())
        
        for step in range(n_steps):
            print(sess.run(b))
    
            global_step.assign_add(1).eval()
            print(global_step.eval())
    
            saver.save(sess, './my_test_model', global_step=global_step)

**Script that restores checkpoint:**

    import tensorflow as tf
    
    from tensorflow.python.tools.inspect_checkpoint import print_tensors_in_checkpoint_file
    
    # List ALL tensors.
    print_tensors_in_checkpoint_file(tf.train.latest_checkpoint('./'), all_tensors=True, tensor_name='')
    
    tf.reset_default_graph()
    
    a = tf.get_variable('a', shape=[])
    b = tf.get_variable('b', shape=[])
    global_step = tf.get_variable('global_step', shape=[])
    
    saver = tf.train.Saver()
    
    with tf.Session() as sess:
            
        ckpt = tf.train.latest_checkpoint('./')
        if ckpt:
            print(ckpt)
        
            saver.restore(sess, ckpt)
        
        else:
            print('Nothing restored')
        
        print(a.eval())
        print(b.eval())
        print(global_step.eval())

**Output**
```
tensor_name:  a
3.0
tensor_name:  b
20.0
tensor_name:  global_step
5
./my_test_model-5
INFO:tensorflow:Restoring parameters from ./my_test_model-5
3.0
20.0
7e-45
```

**Comment by tensorflowbutler** [comment]
Nagging Assignee @drpngx: It has been 15 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by drpngx** [comment]
Yes, some extra validation sounds like a good idea.

**Comment by facaiy** [comment]
@drpngx I create a PR #21078 to fix the issue, could you take a look? Thanks.
